jsk_topic_tools/src/synchronized_throttle_nodelet.cpp

```diff
@@ -0,0 +1,418 @@
+                            *sub_[4], null_, null_, null_);
+        break;
+      case 6:
+        sync_->connectInput(*sub_[0], *sub_[1], *sub_[2], *sub_[3],
+                            *sub_[4], *sub_[5], null_, null_);
+        break;
+      case 7:
+        sync_->connectInput(*sub_[0], *sub_[1], *sub_[2], *sub_[3],
+                            *sub_[4], *sub_[5], *sub_[6], null_);
+        break;
+      case 8:
+        sync_->connectInput(*sub_[0], *sub_[1], *sub_[2], *sub_[3],
+                            *sub_[4], *sub_[5], *sub_[6], *sub_[7]);
+        break;
+      default:
+        NODELET_FATAL("Unhandled error");
+        return;
+    }
+    sync_->registerCallback(
+        boost::bind(&SynchronizedThrottle::inputCallback, this,
+                    _1, _2, _3, _4, _5, _6, _7, _8));
+  }
+}
+
+void SynchronizedThrottle::unsubscribe()
+{
+  NODELET_DEBUG("unsubscribe");
+
+  for (size_t i = 0; i < sub_.size(); ++i)
+  {
+    sub_[i]->unsubscribe();
+  }
+}
+
+bool SynchronizedThrottle::isSubscribed() const
+{
+  return subscribed_;
+}
+
+void SynchronizedThrottle::checkCallback(
+    const topic_tools::ShapeShifterStamped::ConstPtr& msg,
+    const size_t index)
+{
+  boost::mutex::scoped_lock lock(mutex_);
+
+  NODELET_DEBUG_STREAM("check callback: " << index);
+  NODELET_DEBUG_STREAM(" name: " << input_topics_[index]);
+  NODELET_DEBUG_STREAM(" type: " << msg->getDataType());
+  NODELET_DEBUG_STREAM(" md5: " << msg->getMD5Sum());
+
+  check_sub_[index].shutdown();
+
+  // advertise
+  ros::SubscriberStatusCallback connect_cb
+      = boost::bind(&SynchronizedThrottle::connectCb, this);
+  ros::SubscriberStatusCallback disconnect_cb
+      = boost::bind(&SynchronizedThrottle::disconnectCb, this);
+  std::string output_topic = input_topics_[index] + "/" + suffix_;
+  ros::AdvertiseOptions options(
+      output_topic, 1,
+      msg->getMD5Sum(), msg->getDataType(), msg->getMessageDefinition(),
+      connect_cb, disconnect_cb);
+  options.latch = false;
+  pub_[index] = pnh_->advertise(options);
+
+  // check if all are advertised
+  bool all_advertised = true;
+  for (size_t i = 0; i < pub_.size(); ++i)
+  {
+    if (!pub_[i]) all_advertised = false;
+  }
+  if (all_advertised)
+  {
+    NODELET_DEBUG("All Advertised");
+    advertised_ = true;
+    if (!subscribed_)
+    {
+      for (size_t i = 0; i < pub_.size(); ++i)
+      {
+        if (pub_[i].getNumSubscribers() > 0) {
+          subscribe();
+          subscribed_ = true;
+          break;
+        }
+      }
+    }
+  }
+}
+
+void SynchronizedThrottle::connectCb()
+{
+  boost::mutex::scoped_lock lock(mutex_);
+
+  NODELET_DEBUG("connectCb");
+
+  if (advertised_ && !subscribed_)
+  {
+    for (size_t i = 0; i < pub_.size(); ++i)
+    {
+      if (pub_[i].getNumSubscribers() > 0) {
+        subscribe();
+        subscribed_ = true;
+        break;
+      }
+    }
+  }
+}
+
+void SynchronizedThrottle::disconnectCb()
```

Comment:
Hmm, I just found that only if `approximate_sync` is enabled, unsubscription is delayed for 10~20 seconds and I have no idea about the behavior. (the function call `disconnectCb` itself is delayed)
But IMO, it is a minor issue and it is assumed that this nodelet is used mostly with exact synchronization, so still I'd like to be merged now.